### ISSUE TITLE
[sql] Can't save binary data with `execute` (SQLite)

### ISSUE THREAD

**Issue by joakim** [tauri-apps/plugins-workspace#105]
First of all, thanks for your work on Tauri and essential plugins like this one!

I'm having trouble inserting `Uint8Array` data into `BLOB` columns in SQLite. Instead of being inserted as binary data, it's inserted as a stringified JSON object.

My main use case is a `uuid` column storing the 128-bit binary representation of a UUID.

```sql
CREATE TABLE node (
    uuid BLOB PRIMARY KEY
    -- ...
)
```

```typescript
const uuid = new Uint8Array(16)  // this would be a generated 128-bit UUID
const result = await db.execute('INSERT INTO node (uuid) VALUES ($1)', [uuid])
```

What's inserted into `node.uuid`:

```js
'{"0":0,"1":0,"10":0,"11":0,"12":0,"13":0,"14":0,"15":0,"2":0,"3":0,"4":0,"5":0,"6":0,"7":0,"8":0,"9":0}'
```

If I instead save the `ArrayBuffer` itself, I get an empty object:

```typescript
const uuid = new Uint8Array(16).buffer  // its underlying ArrayBuffer
const result = await db.execute('INSERT INTO node (uuid) VALUES ($1)', [uuid])
```

```js
'{}'
```

Am I doing something wrong, or does the binary data simply not survive serialization when sent from JS to Rust?

If it's the latter, maybe [serde_binary](https://docs.rs/serde-binary/latest/serde_binary/) can be of help?

**Comment by FabianLars** [comment]
uint8array is not an actual array but just a plain JS object which the ipc converts to a json object. what's been necessary so far (and used for the core writeBinaryFile api for examples, is to convert it to an actual array before sending it back -> `Array.from(uuid)`

so i guess we should at that to the plugin too

**Comment by joakim** [comment]
Yes, I think the API should support TypedArrays too for representing binary data.

As long it's saved as binary data and returned as binary data by sqlx, `Array.from` sounds good :)

**Comment by patrick91** [comment]
I have this working using something like:

```
      Object.values(JSON.parse(snapshotTextFromDb))
```

but I wonder if we should not use JSON to send data to Rust, if I understood correct the new IPC system in Tauri 2 supports sending binary data, should we switch to that? 

My issue at the moment is that I'm sending a ~200kb binary document and it's quite slow (and it's also hanging the main thread)

**Comment by liuzhch1** [comment]
hi, I'd like to fix this issue. Here is what I found.
https://github.com/tauri-apps/tauri/pull/10797 has fixed the serialization of Uint8Array and ArrayBuffer.
Currently they are sent as JSON array. Then they will be transformed to string when binding to binary column in rust side. Like `'[0,0,0,0]'`
> new IPC system in Tauri 2 supports sending binary data, should we switch to that?

Yeah tauri support `Raw(Vec<u8>)` as `InvokeBody`, however sql plugin pass a [object](https://github.com/tauri-apps/plugins-workspace/blob/fa601e8754d09634f87db642a2a7eee5796e7b78/plugins/sql/guest-js/index.ts#L111-L115) to ipc, which isn't supported to be sent as [binary](https://github.com/tauri-apps/tauri/blob/dd7e59a4956efef291e17295ba08841e9d74e505/crates/tauri/scripts/process-ipc-message-fn.js#L8-L17)(`application/octet-stream`)

A solution I've come up with is using a special identifier for binary data, and encode them as base64.
It looks something like:
<details>

```ts
// plugins/sql/guest-js/index.ts
async execute(query: string, bindValues?: unknown[]): Promise<QueryResult> {
 const values = (bindValues ?? []).map((v) => {
   if (v instanceof Uint8Array) {
     return { __tauri_bytes__: v.toBase64() }
   } else if (v instanceof ArrayBuffer) {
     return { __tauri_bytes__: new Uint8Array(v).toBase64() }
   }
   return v
 })
 ...
}
async select<T>(query: string, bindValues?: unknown[]): Promise<T> {
 const result = await invoke<T>('plugin:sql|select', {
 let result = await invoke<T>('plugin:sql|select', {
    db: this.path,
    query,
    values: bindValues ?? []
  })
 if (Array.isArray(result))
   result = result.map((v) => {
     if (!(typeof v === 'object' && v !== null)) return v as T
     return Object.fromEntries(
       Object.entries(v as Record<string, unknown>).map(([k, v]) => {
         if (!(typeof v === 'object' && v !== null)) return [k, v]
         return '__tauri_bytes__' in v
           ? [k, Uint8Array.fromBase64(v['__tauri_bytes__'])]
           : [k, v]
       })
     )
   })
  return result
}
```
Rust side
```diff
plugins/sql/src/decode/sqlite.rs
"BLOB" => {
if let Ok(v) = v.to_owned().try_decode::<Vec<u8>>() {
-  JsonValue::Array(v.into_iter().map(|n| JsonValue::Number(n.into())).collect())
+  use base64::Engine;
+  let data_base64 = base64::engine::general_purpose::STANDARD.encode(v);
+  JsonValue::Object(serde_json::Map::from_iter([(
+      "__tauri_bytes__".to_string(),
+      data_base64.into(),
+  )]))
```
</details>

This implementation require base64 encode and decode. So additional dependencies may required. Browser do have a [Uint8Array.toBase64](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array/toBase64), but it's recently available. So it requires copy paste base64 impl, thirdparty package or polyfill etc.

And I did a quick test on performance, execute INSERT of binary data.

200kb:  50ms current, 20ms using base64
2MB:     500ms current, 50ms using base64

cc @FabianLars. Hi, please correct me if I misunderstood them. And any comments about this solution?

**Comment by FabianLars** [comment]
how about always sending bytes? and converting that via `JSON.parse(Array.from(bytes))` or whatever

wondering if that's faster 🤔

**Comment by liuzhch1** [comment]
Do you mean directly dump the {db,query,bindValues} object into bytes, transform Uint8Array to number[] before dump?
When `execute`, Rust side need to know which field is binary to correctly bind it as Vec<u8> to query.
Same for `select`, Rust side need to tell which field is binary, otherwise JS side don't know it's number[] data or it should be transformed to Uint8Array.

**Comment by FabianLars** [comment]
> Do you mean directly dump the {db,query,bindValues} object into bytes

Yes, the IPC handles command arguments like one json object anyway so we can do the same eg converting it to json then to binary.

> transform Uint8Array to number[] before dump?

yep

> When execute, Rust side need to know which field is binary to correctly bind it as Vec to query.

if we send json-as-binary then serde_json should be able to figure that out itself.

> Same for select, Rust side need to tell which field is binary, otherwise JS side don't know it's number[] data or it should be transformed to Uint8Array.

We don't have to convert it to uint8array - since number[] is the most native type for this we might as well return that as-is and let the user decide whether they want uint8array - we do the same in tauri's invoke for example.

**Comment by liuzhch1** [comment]
> if we send json-as-binary then serde_json should be able to figure that out itself.

vec will be serialized to string when binding to binary field, e.g. Blob of sqlite. It takes significantly more space? 👀

**Comment by FabianLars** [comment]
base64 increases the size of whatever i encodes by 33% - if we serialize the whole invoke payload to binary ourselves (talking about `Raw(Vec<u8>)` if that wasn't clear) then it shouldn't increase by that much and hopefully be a bit faster. json does support number arrays which are fairly easy to get from blobs as well. Sorry if i miss something else here that would bloat the filesize, i didn't think about IPC stuff for a while...

In theory we could also use MessagePack or BSON or whatever if we wanted to but that's probably overkill if not done across more plugins and/or in tauri itself.

btw, this issue wasn't exactly about performance issues, but rather just the missing Array.from uses so in theory it may be fixed now with https://github.com/tauri-apps/tauri/pull/10797 (except for Blobs if we want to support that)

> Currently they are sent as JSON array. Then they will be transformed to string when binding to binary column in rust side. Like '[0,0,0,0]'

that doesn't sound right but sounds fixable?

**Comment by liuzhch1** [comment]
Yeah this issue isn't about performance, so let's put it aside for now.

> that doesn't sound right but sounds fixable?

The problem arises because when `value` is [bind](https://github.com/tauri-apps/plugins-workspace/blob/fa601e8754d09634f87db642a2a7eee5796e7b78/plugins/sql/src/wrapper.rs#L163), the `value` is `Vec<Number>`(a `JsonValue`). It seems being converted into a string and passed to SQLite, which will write the string into a BLOB field. The result will be `'[0,0,0,0]'`.
So the plugin didn't correctly handle BLOB field without *data type hint*. (same for query BLOB data, the query result is string `'[0,0,0,0]'` instead of JS array `[0,0,0,0]`)
Rust side may need some hint to identify binary fields, and correctly bind to BLOB column.
[reproduction](https://github.com/liuzhch1/temp-tauri-sql-blob)

The docs of `bind` states:
> There is no validation that the value is of the type expected by the query. Most SQL flavors will perform type coercion (Postgres will return a database error).

Sqlite seems happy to store string into BLOB field.

**Comment by FabianLars** [comment]
ugh that sucks 🙃 we could however inject some magic bytes in front of the array if we wanted to (eg `__tauri_bytes__` as bytes) or send another arg with more information on the fields

not sure if that's worth it though so not forcing you to try that

**Comment by liuzhch1** [comment]
> eg `__tauri_bytes__` as bytes

What do you mean by 'bytes'? Currently it sends as json array(converted from Uint8Array). Add string to front? `['__tauri__bytes__',0,0,0,0]`? Its alike wrapping number[] in a object as proposed above. (without considering BSON or MessagePack you mentioned since they are performance scope, I haven't used them before)

Add additional arg is also ok, I guess? Add information like `typeHint: [[idx, 'binary']]` the idx bindValue is a binary

I'd like to fix this issue, so yeah I'd like to try

**Comment by FabianLars** [comment]
> What do you mean by 'bytes'?

something like
```ts
let enc = new TextEncoder();
let bytesMarker = Array.from(enc.encode("__tauri__bytes__"));
let payload = [...bytesMarker, 0, 0, 0, 0]
```
but hardcoded of course to safe a nanosecond computetime 😂 

But an additional arg is probably a cleaner approach albeit potentially slower cause it's more we have to deserialize on the rust side 🤷

**Comment by joakim** [comment]
Thanks for looking into this!

In case a serialization format is something you want to consider, it seems that Apache Avro does well in benchmarks. Faster than JSON even.

https://github.com/Adelost/javascript-serialization-benchmark?tab=readme-ov-file#result-final  
https://adamfaulkner.github.io/binary_formats_are_better_than_json_in_browsers.html

Edit: I was too quick on the trigger, Avro is not schema-less. That said, schema-less formats like MessagePack, CBOR and BSER are close to JSON in performance and smaller in size.

**Comment by FabianLars** [comment]
Let's say i'm open to the idea 🤷 i'd probably wait and make this something tauri-wide but i wouldn't mind testing it in a plugin or two upfront. whatever you prefer really. sticking with json would be the less breaking change i guess but breaking nonetheless
